smart-wallets/max-limit-smart-wallet/max-limit-smart-wallet.pact

```diff
@@ -0,0 +1,125 @@
+    )
+
+  (deftable wallet-table:{wallet})
+
+  ; --------------------------------------------------------------------------
+  ; Module Guards
+
+  (defun wallet-module-guard:guard()
+    (create-module-guard 'wallet-guard))
+
+  ; --------------------------------------------------------------------------
+  ; Capabilities
+
+  (defcap CUSTODIAL:bool (account)
+    @doc "Lookup and enforce custodial guards associated with an account"
+    (with-read wallet-table account {
+        "custodial-guard":= guard
+      }
+      (enforce-guard guard)))
+
+  (defcap NON_CUSTODIAL:bool (account)
+    @doc "Lookup and enforce non-custodial guards associated with an account"
+    (with-read wallet-table account {
+        "non-custodial-guard":= guard
+      }
+      (enforce-guard guard)))
+
+  (defcap AMOUNT_CHECK:bool (account:string amount:decimal)
+    @doc "Lookup and enforce that the tx amount is in the range of allowed custodial tx amount"
+
+    (with-read wallet-table account {
+      "max-custodial-amount":= max
+      }
+      (enforce (> max amount) "Amount is bigger than maximum")))
+
+  ; --------------------------------------------------------------------------
+  ; Smart Wallet Functions
+
+  (defun create-wallet-account:string (account:string custodial-guard:guard non-custodial-guard:guard max-custodial-amount:decimal)
+    @doc "Create a coin account with ACCOUNT and WALLET_MODULE_GUARD in coin-table,     \
+    \and create a wallet account with ACCOUNT, CUSTODIAL_GUARD, and NON_CUSTODIAL_GUARD,\
+    \ and MAX_CUSTODIAL_AMOUNT "
+
+    (enforce-guard custodial-guard)
+    (enforce-guard non-custodial-guard)
+
+    ;;Create Account in Kadena Coin Table with Wallet Module Guard.
+    ;;This Prevents other contracts from tx functions of accounts created by the wallet.
+    (create-account account (wallet-module-guard))
+    (insert wallet-table account {
+        "custodial-guard": custodial-guard,
+        "non-custodial-guard": non-custodial-guard,
+        "max-custodial-amount": max-custodial-amount
+       }))
+
+  (defun custodial-transfer:string (sender:string receiver:string amount:decimal)
+    @doc "This function enforces that tx amount is smaller than SENDER's    \
+    \MAX_CUSTODIAL_AMOUNT, and then enforces the SENDER's CUSTODIAL_GUARD. \
+    \ If the conditions qualify, transfers AMOUNT of Kadena Coin from SENDER \
+    \account to RECEIVER account."
+
+    (with-capability (AMOUNT_CHECK sender amount)
+      (with-capability (CUSTODIAL sender)
+        (transfer sender receiver amount))))
+
+  (defun non-custodial-transfer:string (sender:string receiver:string amount:decimal)
+    @doc "This function enforces SENDER's NON_CUSTODIAL_GUARD and transfers \
+    \ AMOUNT of Kadena coins from SENDER account to RECEIVER account without \
+    \limit."
+
+    (with-capability (NON_CUSTODIAL sender)
+      (transfer sender receiver amount)))
+
+  (defun update-max-custodial-amount:string (account:string amount:decimal)
```

Comment:
We should probably make another variant that enforces a limit on the amount as well as a limit on the number of withdrawals that can be done in a particular period of time.  Otherwise an attacker can just issue a bunch of small transactions.